README
 $Id: README,v 1.4.2.2 2013/04/16 09:08:26 ak Exp $
-----------------------------------------------------------------------
 Copyright (C) 2009-2013 Cubicroot Co. Ltd. ALL RIGHTS RESERVED.
 Developed by Cubicroot Co. Ltd. (http://cubicroot.jp/)
------------------------------------------------------------------------
 bounceHammer: An open source software mainly developed for handling 
               email bounces on UNIX/Linux server.

 Other information is available at the following sites:

  - http://sourceforge.jp/projects/bouncehammer/
  - http://www.facebook.com/bouncehammer
  - http://twitter.com/bouncehammer
  - http://bouncehammer.jp/
------------------------------------------------------------------------

ARTISTIC -------------- Artistic License
AUTHORS --------------- Author of bounceHammer
GNU-GPL --------------- GNU General Public License version 2
INSTALL --------------- How to install bounceHammer
Implement.PL ---------- Check -T, -B operation on file handle
KNOWNBUGS ------------- Known bugs in bounceHammer
LICENSE --------------- About license
Makefile.am ----------- Makefile
Makefile.in ----------- Makefile
Modules.PL* ----------- Check depended perl modules
README ---------------- This file
RELEASE_NOTES --------- Release history of bounceHammer(English)
RELEASE_NOTES.JA ------ Release history of bounceHammer(Japanese)
SUPPORT --------------- About bounceHammer commercial support
acinclude.m4 ---------- For autoconf
aclocal.m4 ------------
config.guess ----------
config.sub ------------
configure ------------- configure script
configure.ac ----------
diff/ ----------------- Patches for a certain environment
examples/ ------------- Examples for make test
install-sh ------------
missing ---------------
src/ ------------------ Source code
t/ -------------------- Test cases for make test
template/ ------------- WebUI template files



RELEASE_NOTES
 $Id: RELEASE_NOTES,v 1.76.2.26 2013/04/17 03:08:42 ak Exp $
 -Id: RELEASE_NOTES,v 1.12 2009/12/01 10:35:23 ak Exp -
 -Id: RELEASE_NOTES,v 1.34 2009/08/27 05:09:04 ak Exp -
-----------------------------------------------------------------------
 ##                                      ##  ##                                     
 ##      ####  ##  ## #####   #### ####  ##  ##  ####  ##  ## ##  ##  ####  #####   
 #####  ##  ## ##  ## ##  ## ##   ##  ## ######     ## ###### ###### ##  ## ##  ##  
 ##  ## ##  ## ##  ## ##  ## ##   ###### ##  ##  ##### ###### ###### ###### ##      
 ##  ## ##  ## ##  ## ##  ## ##   ##     ##  ## ##  ## ##  ## ##  ## ##     ##      
 #####   ####   ##### ##  ##  #### ####  ##  ##  ##### ##  ## ##  ##  ####  ##      

                   BOUNCEHAMMER/RELEASE_NOTES
-----------------------------------------------------------------------
 Copyright (C) 2009-2011 Cubicroot Co. Ltd. ALL RIGHTS RESERVED.
 Developed by Cubicroot Co. Ltd. (http://cubicroot.jp/)
-----------------------------------------------------------------------
 This listing show the release notes of "bounceHammer",  an open source
 software mainly developed for handling email bounces on UNIX server.
 The list shows the version of bounceHammer engine, the version of con-
 figuration file of bounceHammer, the date of release, and a summary of
 the changes in that release. More information about bounceHammer is a-
 vailable at our site: http://bouncehammer.jp/ .

 Other information is available at the following sites:

  - http://sourceforge.jp/projects/bouncehammer/
  - http://twitter.com/bouncehammer
  - http://www.facebook.com/bouncehammer
-----------------------------------------------------------------------

2.7.10/2.7.0	2013/04/19

  New functionality:
    * (EXPERIMENTAL IMPLEMENTATION): mailboxparser can parse  a bounced
      email from OpenSMTPD (Tested with OpenSMTPD 5.2 and 5.3.1).

  Assorted improvements and code cleanup:
    * Output data (json, and yaml) created by bin/mailboxparser command
      includes the new sub field "listid" in the description field. The
      listid field is content of "List-Id" header generated by mailling
      list software. Thanks to @masaki_ipodrepo and @revsystem .

    * Email domain name of Nokia Mail & Nokia Chat; formerly known as
      Ovi by Nokia, @nokiamail.com is classified into provider 'nokia',
      is classified into hostgroup 'smartphone'.

    * Support some domain names of outlook.com.

    * Analytical accuracy of mailboxparser improvement for Gmail.

    * Pick the first email address from Reply-To: header  if there are
      2 or more email addresses in the header.

  Fixed bugs:
    * Fixed an issue while processing a Diagnostic-Code: header across
      multiple lines. Thanks to Arie Valstar.

  New files:
    * lib/Kanadzuchi/MTA/OpenSMTPD.pm (Experimental)
    * t/042_mta-opensmtpd.t (Experimental)


-----------------------------------------------------------------------
2.7.9/2.7.0	2012/10/02

  Assorted improvements and code cleanup:
    * @outlook.com is classified into Webmail/microsoft .

    * @india.com and other India.com's domain names are classified into
      Webmail/india.com.

    * Improved classification of some major webmail and mobile domains:
      aol.*, mac.me, hotmail.*, yahoo.*, and *.vodafone.*.
 
  Fixed bugs:
    * Version requirement of "Time::Piece" is 1.20  for running command
      './configure'. Thanks to @memememomo .

    * Improved module's version checks in ./Modules.PL(use version).

    * Fixed an issue that Internet Explorer or Google Chrome gets a 500
      error on WebUI. The issue might be occurred by "Content Sniffing"
      feature, so we have added "X-Content-Type-Options: nosniff".
      Thanks to Donald DeWulf.


-----------------------------------------------------------------------
2.7.8/2.7.0	2012/04/09

  New functionality:
    * (EXPERIMENTAL IMPLEMENTATION): bin/mailboxparser detects some of
      FBL: Feedback loop message as a bounced email. Thanks to Gilles.

  Fixed bugs:
    * Fixed an issue that  src/cgi/Makefile uses "/" as a temporary di-
      rectory when the "make" command is executed. Thanks to Hasegawa.

    * Fixed an issue that Kanadzuchi::Address->canonify fails when the
      email header does not contain any space character like follwoing:
      "=?ISO-2022-JP?B?.....?="<user@example.jp> . Thanks to Kanazawa.

  New files:
    * lib/Kanadzuchi/MTA/FeedbackLoop.pm (Experimental)
    * t/041_mta-feedbackloop.t (Experimental)


-----------------------------------------------------------------------
2.7.7/2.7.0	2011/11/14

  New functionality:
    * New command line option for bin/mailboxparser, --save-failed-mesg
      is for save messages which are failed to parse.

  Assorted improvements and code cleanup:
    * Update for commercial MTA modules.
    * Improved analytical accuracy for bounce messages via Exim, qmail,
      Sendmail, Postfix, and au by KDDI. Thanks to Toyokawa.


-----------------------------------------------------------------------
2.7.6/2.7.0	2011/10/03

  Assorted improvements and code cleanup:
    * @m.facebook.com is classified into Webmail/facebook .
    * @verizon.net is classified into Smartphone/verizon.

  Fixed bugs:
    * Fix the 3rd argument of AC_PATH_PROG() macro in ./configure .
    * Fixed an issue that  bin/mailboxparser stops  with the error mes-
      sage: ``Invalid mailbox list: ...'' if any ESC(0x1b) character is
      contained in a phrase part of a email header. Thanks to Toyokawa.



src/lib/Kanadzuchi/Mbox.pm
# $Id: Mbox.pm,v 1.28.2.14 2013/04/15 04:20:52 ak Exp $
# -Id: Parser.pm,v 1.10 2009/12/26 19:40:12 ak Exp -
# -Id: Parser.pm,v 1.1 2009/08/29 08:50:27 ak Exp -
# -Id: Parser.pm,v 1.4 2009/07/31 09:03:53 ak Exp -
# Copyright (C) 2009-2013 Cubicroot Co. Ltd.
# Kanadzuchi::
                             
 ##  ## ##                   
 ###### ##      #### ##  ##  
 ###### #####  ##  ## ####   
 ##  ## ##  ## ##  ##  ##    
 ##  ## ##  ## ##  ## ####   
 ##  ## #####   #### ##  ##  
package Kanadzuchi::Mbox;

# See also
#  * http://en.wikipedia.org/wiki/Comparison_of_mail_servers
#  * http://en.wikipedia.org/wiki/List_of_mail_servers

#  ____ ____ ____ ____ ____ ____ ____ ____ ____ 
# ||L |||i |||b |||r |||a |||r |||i |||e |||s ||
# ||__|||__|||__|||__|||__|||__|||__|||__|||__||
# |/__\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|
#
use base 'Class::Accessor::Fast::XS';
use strict;
use warnings;
use Perl6::Slurp;
use JSON::Syck;
use Encode;
use Kanadzuchi::MTA::Sendmail;
use Kanadzuchi::MTA::Postfix;
use Kanadzuchi::MTA::qmail;
use Kanadzuchi::MTA::Exim;
use Kanadzuchi::MTA::Courier;
use Kanadzuchi::MTA::OpenSMTPD;
use Kanadzuchi::MTA::FeedbackLoop;

#  ____ ____ ____ ____ ____ ____ ____ ____ ____ 
# ||A |||c |||c |||e |||s |||s |||o |||r |||s ||
# ||__|||__|||__|||__|||__|||__|||__|||__|||__||
# |/__\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|
#
__PACKAGE__->mk_accessors(
	'file',		# (String) File name to parse
	'greed',	# (Integer) Flag, 1 is greedily parse
	'emails',	# (Ref->Array) eMails, Raw test data
	'nmails',	# (Interger) The number of eMails
	'messages',	# (Ref->Array) Messages(Ref->Hash)
	'nmesgs',	# (Integer) The number of parsed messages
);

#  ____ ____ ____ ____ ____ ____ ____ ____ 
# ||C |||o |||n |||s |||t |||a |||n |||t ||
# ||__|||__|||__|||__|||__|||__|||__|||__||
# |/__\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|
#
sub ENDOF() { qq(\n__THE_END_OF_THE_EMAIL__\n); }
my $TransferAgents = __PACKAGE__->postulat();
my $MostFamousMTAs = [ 
	'Sendmail', 'Postfix', 'qmail', 'Exim', 'Courier',
	'OpenSMTPD', 'FeedbackLoop'
];

#  ____ ____ ____ ____ ____ _________ ____ ____ ____ ____ ____ ____ ____ 
# ||C |||l |||a |||s |||s |||       |||M |||e |||t |||h |||o |||d |||s ||
# ||__|||__|||__|||__|||__|||_______|||__|||__|||__|||__|||__|||__|||__||
# |/__\|/__\|/__\|/__\|/__\|/_______\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|
#
sub new
{
	# +-+-+-+
	# |n|e|w|
	# +-+-+-+
	#
	# @Description	Wrapper method of new()
	# @Param
	# @Return	Kanadzuchi::Mbox Object
	my $class = shift;
	my $argvs = { @_ };

	DEFAULT_VALUES: {
		$argvs->{'greed'} = 0 unless $argvs->{'greed'};
		$argvs->{'nmails'} = 0;
		$argvs->{'emails'} = [];
		$argvs->{'nmesgs'} = 0;
		$argvs->{'messages'} = [];
	}

	return $class->SUPER::new( $argvs );
}

sub postulat
{
	# +-+-+-+-+-+-+-+-+
	# |p|o|s|t|u|l|a|t|
	# +-+-+-+-+-+-+-+-+
	#
	# @Description	Require Kanadzuchi::MTA::??::*
	# @Param	<None>
	# @Return	(Ref->Array) Loaded class names
	# @See		etc/avalable-countries
	#
	#     agents: [ 'name-of-mta1', 'name-of-mta2', ... ]
	#       * The key 'agents' does not exist: Load all of modules in Kanadzuchi::
	#         Mbox::<CCTLD or ISO3166>::*.pm
	#       * The key 'agents' is EMPTY: Does NOT load any moduels in Kanadzuchi::
	#         Mbox::<CCTLD or ISO3166>::*.pm
	#       * The key 'agents' has a value: Load only the module of its name in 
	#         Kanadzuch::Mbox::<CCTLD or ISO3166>::<its name>.pm
	my $class = shift;

	# Experimental implementation for the future.
	my $libmboxroot = '__KANADZUCHILIB__/Kanadzuchi/MTA';
	my $additionals = [ 'User', 'Comm' ];
	my $iso3166list = [ 'JP', 'US' ];
	my $iso3166conf = '__KANADZUCHIETC__/available-countries';
	my $countryconf = ( -r $iso3166conf && -s _ && -T _ ) ? JSON::Syck::LoadFile($iso3166conf) : {};
	my $didfileload = keys %$countryconf ? 1 : 0;
	my $listofclass = [];
	my $acclassname = q();

	EACH_COUNTRY: foreach my $code ( @$iso3166list )
	{
		# etc/avalable-countries does not exist, load all of modules in 
		# Kanadzuchi/MTA/??/*.pm
		my $directory = $libmboxroot.'/'.$code;
		my $mtaoption = [];	# Require files in this array

		# The directory does not exist or is not readable, or is not executable.
		next(EACH_COUNTRY) unless( -d $directory && -r _ && -x _ );

		if( $didfileload && exists( $countryconf->{ lc $code }->{'agents'} ) )
		{
			# The key 'agents' exists, check the value of the key
			$mtaoption = $countryconf->{ lc $code }->{'agents'};

			# agents: []
			#   * The key 'agents' is EMPTY: Does NOT load any moduels in 
			#     Kanadzuchi::MTA::<CCTLD or ISO3166>::*.pm
			next(EACH_COUNTRY) unless( scalar @$mtaoption );

			# agents: [ 'something', ... ]
			#   * The key 'agents' has a value: Load only the module of its
			#     name in Kanadzuch::MTA::<CCTLD or ISO3166>::<its name>.pm
			map { $_ = lc $_.'.pm' } @$mtaoption;
		}

		opendir( my $dh, $directory );
		READDIR: while( my $de = readdir($dh) )
		{
			my $fp = $directory.'/'.$de;

			# the file is not *.pm, nor regular file, nor readable
			next(READDIR) if( $fp !~ m{[.]pm\z} || ! -f $fp || ! -r _ );
			next(READDIR) if( scalar @$mtaoption && ! grep { lc($de) eq $_ } @$mtaoption );

			$acclassname  = 'Kanadzuchi::MTA::'.$code.'::'.$de;
			$acclassname =~ s{[.]pm\z}{};

			eval { require $fp; };
			push @$listofclass, $acclassname unless $@;
		}
		closedir $dh;

	} # End of foreach(EACH_COUNTRY)

	ADDITIONALS: foreach my $amod ( @$additionals )
	{
		# Load user-defined modules and commercial MTA modules
		my $directory = $libmboxroot.'/'.$amod;

		next unless( -d $directory && -r _ && -x _ );
		opendir( my $dh, $directory );

		READAM: while( my $de = readdir($dh) )
		{
			my $fp = $directory.'/'.$de;

			# the file is not *.pm, nor regular file, nor readable
			next(READAM) if( $fp !~ m{[.]pm\z} || ! -f $fp || ! -r _ );

			$acclassname  = 'Kanadzuchi::MTA::'.$amod.'::'.$de;
			$acclassname =~ s{[.]pm\z}{};

			eval { require $fp; };
			push @$listofclass, $acclassname unless $@;
		}

		closedir $dh;

	} # End of foreach(ADDITIONALS)

	return $listofclass;
}

sub breakit
{
	# +-+-+-+-+-+-+-+
	# |b|r|e|a|k|i|t|
	# +-+-+-+-+-+-+-+
	#
	# @Description	Break the header of message and return its body
	# @Param <ref>	(Ref->Hash) Message entity.
	# @Param <ref>	(Ref->String) Message body
	# @Return	(String) Message body or empty string
	my $packagename = shift;
	my $thismessage = shift || return q();
	my $thebodypart = shift || return q();
	my $theheadpart = $thismessage->{'head'};

	# For Code Refactoring IN THE FUTURE.
	# my $contenttype = [	
	#	qr{\Amultipart/(?:report|mixed)},
	#	qr{\Amessage/(?:delivery-status|rfc822)},
	#	qr{\Atext/rfc822-headers},
	# ];

	# Check whether or not the message is a bounce mail.
	#  _____             _     _____                                _          _ 
	# |  ___|_      ____| |_  |  ___|__  _ ____      ____ _ _ __ __| | ___  __| |
	# | |_  \ \ /\ / / _` (_) | |_ / _ \| '__\ \ /\ / / _` | '__/ _` |/ _ \/ _` |
	# |  _|  \ V  V / (_| |_  |  _| (_) | |   \ V  V / (_| | | | (_| |  __/ (_| |
	# |_|     \_/\_/ \__,_(_) |_|  \___/|_|    \_/\_/ \__,_|_|  \__,_|\___|\__,_|
	#                                                                            
	# Pre-Process eMail body if it is a forwarded bounce message.
	#  Get forwarded text if a subject begins from 'fwd:' or 'fw:'
	if( lc( $theheadpart->{'subject'} ) =~ m{\A\s*fwd?:} )
	{
		# Break quoted strings, quote symbols(>)
		$$thebodypart =~ s{\A.+?[>]}{>}s;
		$$thebodypart =~ s{^[>]+[ ]}{}gm;
		$$thebodypart =~ s{^[>]$}{}gm;
	}

	#  ____  _                  _               _   _____                          _   
	# / ___|| |_ __ _ _ __   __| | __ _ _ __ __| | |  ___|__  _ __ _ __ ___   __ _| |_ 
	# \___ \| __/ _` | '_ \ / _` |/ _` | '__/ _` | | |_ / _ \| '__| '_ ` _ \ / _` | __|
	#  ___) | || (_| | | | | (_| | (_| | | | (_| | |  _| (_) | |  | | | | | | (_| | |_ 
	# |____/ \__\__,_|_| |_|\__,_|\__,_|_|  \__,_| |_|  \___/|_|  |_| |_| |_|\__,_|\__|
	#                                                                                  
	# Pre-Process eMail headers of standard bounce message
	#return $$thebodypart if( $theheadpart->{'content-type'} && 
	#			grep { $theheadpart->{'content-type'} =~ $_ } @$contenttype );
	#
	my $parserclass = q();		# (String) Package|Class name
	my $pseudofield = q();		# (String) Pseudo headers
	my $agentmodule = q();		# (String) Agent class name
	my $isforwarded = 0;		# (Integer) Is forwarded message

	# Most famous MTAs
	foreach my $mta ( @$MostFamousMTAs )
	{
		$agentmodule  = 'Kanadzuchi::MTA::'.$mta;
		$pseudofield .= $agentmodule->reperit( $theheadpart, $thebodypart );
		last if $pseudofield;
	}

	# Optionals
	unless( $pseudofield )
	{
		foreach my $mod ( @$TransferAgents )
		{
			$pseudofield .= $mod->reperit( $theheadpart, $thebodypart );
			last if $pseudofield;
		}

		# Fallback
		unless( $pseudofield )
		{
			require Kanadzuchi::MTA::Fallback;
			$pseudofield .= Kanadzuchi::MTA::Fallback->reperit( $theheadpart, $thebodypart );

			unless( $pseudofield )
			{
				if( $$thebodypart =~ m{^[Ss]tatus: [45][.][0-7][.]\d+(.*)}m )
				{
					unless( $$thebodypart =~ m{^[Dd]iagnostic-[Cc]ode: }m )
					{
						$pseudofield .= 'X-SMTP-Diagnosis: '.$1.qq(\n);
					}
					$pseudofield .= Kanadzuchi::MTA->xsmtpcommand('DATA');
					$pseudofield .= Kanadzuchi::MTA->xsmtpagent('unknown');
				}
			}
		}
	}
	return $pseudofield.$$thebodypart;
}

#  ____ ____ ____ ____ ____ ____ ____ ____ _________ ____ ____ ____ ____ ____ ____ ____ 
# ||I |||n |||s |||t |||a |||n |||c |||e |||       |||M |||e |||t |||h |||o |||d |||s ||
# ||__|||__|||__|||__|||__|||__|||__|||__|||_______|||__|||__|||__|||__|||__|||__|||__||
# |/__\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|/_______\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|
#
sub slurpit
{
	# +-+-+-+-+-+-+-+
	# |s|l|u|r|p|i|t|
	# +-+-+-+-+-+-+-+
	#
	# @Description	Slurp the email
	# @Param	<None>
	# @Return	(Integer) n = The number of slurped emails
	my $self = shift;
	my $file = defined $self->{'file'} ? $self->{'file'} : \*STDIN;

	unless( ref($file) eq q|SCALAR| )
	{
		return 0 if( $file =~ m{[\n\r]} || $file =~ m{[\x00-\x1f\x7f]} );
		return 0 if(
			! ( -f $file && -T _ && -s _ ) &&
			! ( ref($file) eq q|GLOB| && -T $file )
		);
	}

	$self->{'emails'} = [];

	eval {
		# Slurp the mailbox, Convert from CRLF to LF,
		#
		# mboxes
		#  mboxo	Original mbox implementation
		#  mboxcl	Content-Length field in UNIX From_ line
		#  mboxcl2	From_ in the message body is not quoted.
		#  mboxrd
		#  MMDF		No UNIX From_ line, and blank line at the end of the message body.
		# 		^A^A^A^A(4 Ctrl-As) around the message.
		#  Eudora	No blank line at the end of the message, From_ ???@???
		#  Netscape	From_ -
		@{ $self->{'emails'} } =
			map( { s{\x0d\x0a}{\n}g; y{\x0d\x0a}{\n\n}; q(From ).$_; }
				Perl6::Slurp::slurp( $file,
					{
						'irs' => qr(\nFrom ),
						'chomp' => ENDOF
					}
				)
			);

		if( scalar @{ $self->{'emails'} } )
		{
			$self->{'emails'}->[0] =~ s{\AFrom (.+)\z}{$1}s;
			$self->{'emails'}->[-1] .= ENDOF;
		}
	};

	return 0 if $@;
	$self->{'nmails'} = scalar @{$self->{'emails'}};
	return $self->{'nmails'};
}

sub parseit
{
	# +-+-+-+-+-+-+-+
	# |p|a|r|s|e|i|t|
	# +-+-+-+-+-+-+-+
	#
	# @Description	Parse the email text
	# @Param <ref>	(Integer) Flag for saving failed messages
	# @Param <ref>	(Ref->Code) Callback function
	# @Return	(Integer) n = The number of parsed messages
	my $self = shift;
	my $save = shift || 0;
	my $call = shift || sub {};
	my $ends = ENDOF;
	my $seek = 0;

	my $agentclasses = [ map { 'Kanadzuchi::MTA::'.$_ } @$MostFamousMTAs ];
	my $emailheaders = [ 
		'From', 'To', 'Date', 'Subject', 'Content-Type',
		'Reply-To', 'Message-Id',
	];
	my $agentheaders = [];

	# Load each agent's headers
	map { push @$agentheaders, @{ $_->emailheaders() } } @$agentclasses;
	map { push @$agentheaders, @{ $_->emailheaders() } } @$TransferAgents;


	PARSE_EMAILS: while( my $_email = shift @{ $self->{'emails'} } )
	{
		my $_mail = {};
		my $_from = q();	# 'From_' of UNIX mbox
		my $_head = q();	# Message header as a string
		my $_body = q();	# Message body as a string

		# Callback(), Term::ProgressBar
		$call->();

		if( $_email =~ m{\AFrom[ ](.+?)\n(.+?)\n\n(.+)$ends\z}so )
		{
			$_from = $1;
			$_head = $2;
			$_body = $3;
		}
		elsif( $_email =~ m{\A(.+?)\n\n(.+)$ends\z}so )
		{
			# There is no UNIX From line, Insert pseudo it.
			$_from = q(From MAILER-DAEMON Sun Dec 31 23:59:59 2000);
			$_head = $1;
			$_body = $2;
		}
		else
		{
			next(PARSE_EMAILS);
		}

		# 1. Set the content in UNIX From_ Line
		$_mail->{'from'} = $_from;
		$_mail->{'head'} = { 
			'to' => q(),
			'from' => q(),
			'date' => q(),
			'subject' => q(),
			'reply-to' => q(),
			'received' => [], 
			'message-id' => q(),
			'content-type' => q(),
		};

		# 2. Parse email headers
		my $__continued = 0;	# Flag; Continued from the previous line.
		my $__ehcounter = 0;	# Temporary counter for email headers

		LINES: foreach my $e ( split( qq{\n}, $_head ) )
		{
			HEADERS: foreach my $eh ( @$emailheaders, @$agentheaders )
			{
				next(HEADERS) unless( $e =~ m{\A$eh[:][ ]*}i );
				$_mail->{'head'}->{ lc $eh } = $1 if( $e =~ m/\A${eh}[:][ ]*(.+?)\z/i );
			}

			# Get and concatenate 'Received:' headers
			if( $e =~ m{\AReceived[:][ ]*(.+?)\z}i )
			{
				push( @{ $_mail->{'head'}->{'received'} }, $1 );
				$__continued = 1;
				$__ehcounter = scalar( @{ $_mail->{'head'}->{'received'} } ) - 1;
			}
			elsif( $__continued && $e =~ m{\A[\s\t]+(.+?)\z} )
			{
				# This line is countinued from the previous line.
				next(LINES) unless( scalar @{ $_mail->{'head'}->{'received'} } );
				$_mail->{'head'}->{'received'}->[ $__ehcounter ] .= q( ).$1;
			}
			else
			{
				$__continued = 0;
				$__ehcounter = 0;
			}
		}

		# 3. Rewrite the part of the message body
		$_mail->{'body'} = __PACKAGE__->breakit( $_mail, \$_body );
		$_head = q();

		# 4. Save entire message
		$_mail->{'data'} = $_email if $save;

		# Parse message body
		# Concatenate multiple-lined headers
		next(PARSE_EMAILS) unless $_mail->{'body'};
		$_mail->{'body'} =~ s{^[Ff]rom:[ ]*([^\n\r]+)[\n\r][ \t]+([^\n\r]+)}{From: $1 $2}gm;
		$_mail->{'body'} =~ s{^[Tt]o:[ ]*([^\n\r]+)[\n\r][ \t]+([^\n\r]+)}{To: $1 $2}gm;

		if( $_mail->{'body'} =~ m{^[Dd]iagnostic-[Cc]ode:[\s]*([^\n\r]+)[\n\r]([\s\t]+.+)}ms )
		{
			my $_d1 = 'Diagnostic-Code: '.$1;
			my $_d2 = $2; $_d2 =~ y{\r}{\n};

			foreach my $e ( split( qq(\n), $_d2 ) )
			{
				last unless $e =~ m{\A[\s\t]+};
				$_d1 .= ' '.$e;
			}

			$_d1 =~ y{ }{}s;
			$_mail->{'body'} =~ s{^[Dd]iagnostic-[Cc]ode:[\s]*[^\n\r]+[\n\r][\s\t]+[^\n\r]+}{$_d1}m;
		}

		# Delete non-required headers
		$_mail->{'body'} =~ y{\n}{\n}s;		# Delete blank lines
		$_mail->{'body'} =~ s{^\W.+[\r\n]}{}mg;	# Delete non-email headers

		# Parse greedy
		if( $self->{'greed'} )
		{
			# Addresser
			$_mail->{'body'} =~ s{^[Aa]pparently-[Ff]rom:[ ]*(.+)$}{<<<<: Apparently-From: $1}m;
			$_mail->{'body'} =~ s{^[Rr]esent-[Ff]rom:[ ]*(.+)$}{<<<<: Resent-From: $1}m;
			$_mail->{'body'} =~ s{^[Rr]esent-[Rr]eply-[Tt]o:[ ]*(.+)$}{<<<<: Resent-Reply-To: $1}m;
			$_mail->{'body'} =~ s{^[Rr]esent-[Ss]ender:[ ]*(.+)$}{<<<<: Resent-Sender: $1}m;
			$_mail->{'body'} =~ s{^[Ss]ender:[ ]*(.+)$}{<<<<: Sender: $1}m;

			# Recipient
			$_mail->{'body'} =~ s{^[Aa]pparently-[Tt]o:[ ]*(.+)$}{<<<<: Apparently-To: $1}m;
			$_mail->{'body'} =~ s{^[Ee]nvelope-[Tt]o:[ ]*(.+)$}{<<<<: Envelope-To: $1}m;
			$_mail->{'body'} =~ s{^[Rr]esent-[Tt]o:[ ]*(.+)$}{<<<<: Resent-To: $1}m;
			$_mail->{'body'} =~ s{^[Xx]-[Ee]nvelope-[Tt]o:[ ]*(.+)$}{<<<<: X-Envelope-To: $1}m;

			# Date
			$_mail->{'body'} =~ s{^[Pp]osted:[ ]*(.+)$}{<<<<: Posted: $1}m;
			$_mail->{'body'} =~ s{^[Pp]osted-[Dd]ate:[ ]*(.+)$}{<<<<: Posted-Date: $1}m;
			$_mail->{'body'} =~ s{^[Rr]esent-[Dd]ate:[ ]*(.+)$}{<<<<: Resent-Date: $1}m;
		}

		# Mark required headers
		$_mail->{'body'} =~ s{^[Aa]rrival-[Dd]ate:[ ]*(.+)$}{<<<<: Arrival-Date: $1}m;
		$_mail->{'body'} =~ s{^[Dd]ate:[ ]*(.+)$}{<<<<: Date: $1}gm;
		$_mail->{'body'} =~ s{^[Dd]elivered-[Tt]o:[ ]*(.+)$}{<<<<: Delivered-To: $1}m;
		$_mail->{'body'} =~ s{^[Dd]iagnostic-[Cc]ode:[ ]*(.+)$}{<<<<: Diagnostic-Code: $1}m;
		$_mail->{'body'} =~ s{^[Ee]nvelope-[Ff]rom:[ ]*(.+)$}{<<<<: Envelope-From: $1}m;
		$_mail->{'body'} =~ s{^[Ee]rrors-[Tt]o:[ ]*(.+)([;].+)?$}{<<<<: Errors-To: $1}m;
		$_mail->{'body'} =~ s{^[Ff]inal-[Rr]ecipient:[ ]*[Rr][Ff][Cc]822;[ ]*(.+)$}{<<<<: Final-Recipient: $1}m;
		$_mail->{'body'} =~ s{^[Ff]rom:[ ]*(.+)$}{<<<<: From: $1}gm;
		$_mail->{'body'} =~ s{^[Ff]orward-[Pp]ath:[ ]*(.+)$}{<<<<: Forward-Path: $1}m;
		$_mail->{'body'} =~ s{^[Ll]ast-[Aa]ttempt-[Dd]ate:[ ]*(.+)$}{<<<<: Last-Attempt-Date: $1}m;
		$_mail->{'body'} =~ s{^[Ll]ist-[Ii]d:[ ]*(.+)$}{<<<<: List-Id: $1}m;
		$_mail->{'body'} =~ s{^[Oo]riginal-[Rr]ecipient:[ ]*[Rr][Ff][Cc]822;[ ]*(.+)$}{<<<<: Original-Recipient: $1}m;
		$_mail->{'body'} =~ s{^[Rr]eply-[Tt]o:[ ]*(.+)$}{<<<<: Reply-To: $1}m;
		$_mail->{'body'} =~ s{^[Rr]eturn-[Pp]ath:[ ]*(.+)$}{<<<<: Return-Path: $1}gm;
		$_mail->{'body'} =~ s{^[Rr]everse-[Pp]ath:[ ]*(.+)$}{<<<<: Reverse-Path: $1}m;
		$_mail->{'body'} =~ s{^[Ss]tatus:[ ]*(\d[.]\d[.]\d+).*$}{<<<<: Status: $1}gm;
		$_mail->{'body'} =~ s{^[Tt]o:[ ]*(.+)$}{<<<<: To: $1}m;
		$_mail->{'body'} =~ s{^[Xx]-[Aa]ctual-[Rr]ecipient:[ ]*[Rf][Ff][Cc]822;[ ]*(.+)$}{<<<<: X-Actual-Recipient: $1}m;
		$_mail->{'body'} =~ s{^[Xx]-[Aa]ctual-[Rr]ecipient:[ ]*(.+)$}{<<<<: X-Actual-Recipient: $1}m;
		$_mail->{'body'} =~ s{^[Xx]-[Pp]ostfix-[Ss]ender:[ ]*(.+)$}{<<<<: X-Postfix-Sender: $1}m;
		$_mail->{'body'} =~ s{^[Xx]-[Ee]nvelope-[Ff]rom:[ ]*(.+)$}{<<<<: X-Envelope-From: $1}m;
		$_mail->{'body'} =~ s{^[Xx]-SMTP-Agent:[ ]*(.+)$}{<<<<: X-SMTP-Agent: $1}m;
		$_mail->{'body'} =~ s{^[Xx]-SMTP-Command:[ ]*(.+)$}{<<<<: X-SMTP-Command: $1}m;
		$_mail->{'body'} =~ s{^[Xx]-SMTP-Diagnosis:[ ]*(.+)$}{<<<<: X-SMTP-Diagnosis: $1}m;
		$_mail->{'body'} =~ s{^[Xx]-SMTP-Status:[ ]*(.+)$}{<<<<: X-SMTP-Status: $1}m;
		$_mail->{'body'} =~ s{^[Xx]-SMTP-Recipient:[ ]*(.+)$}{<<<<: X-SMTP-Recipient: $1}m;
		$_mail->{'body'} =~ s{^[Xx]-SMTP-Charset:[ ]*(.+)$}{<<<<: X-SMTP-Charset: $1}m;

		$_mail->{'body'} =~ s{^\w.+[\r\n]}{}gm;			# Delete non-required headers
		$_mail->{'body'} =~ s{^<<<<:\s}{}gm;			# Delete the mark

		# Missing From: header in the body part, See Kanadzuchi/Mail/Bounced.pm line 139 - 164
		unless( $_mail->{'body'} =~ m{
				^(?:From|Return-Path|Reply-To|Errors-To|Reverse-Path
					|X-Postfix-Sender|Envelope-From|X-Envelope-From
					|Resent-From|Sender|Resent-Reply-To|Apparently-From):[ ]}mx ){

			$_mail->{'body'} .= qq(\n) unless $_mail->{'body'} =~ m{\n\z}mx;
			$_mail->{'body'} .= 'From: '.$_mail->{'head'}->{'to'}.qq(\n);
		}

		# Missing Date: header in the body part, See Kanadzuchi/Mail/Bounced.pm line 245 - 253
		unless( $_mail->{'body'} =~ m{
				^(?:Arrival-Date|Last-Attempt-Date|Date
					|Posted-Date|Posted|Resent-Date):[ ]}mx ){

			$_mail->{'body'} .= 'Date: '.$_mail->{'head'}->{'date'}.qq(\n);
		}

		push @{ $self->{'messages'} }, {
				'data' => $_mail->{'data'},
				'from' => $_mail->{'from'},
				'head' => $_mail->{'head'},
				'body' => $_mail->{'body'}, };

		$self->{'nmesgs'}++;

	} # End of while(PARSE_EMAILS)

	$self->{'emails'} = [];
	return $self->{'nmesgs'};
}

1;
__END__
